### PR TITLE
A shared HUD with a round timer

### PR DESCRIPTION
**Pull request by Fueredoriku**
Includes both the visual and functional parts of the round timer.

Things to note:
- If a round timer runs out, players are sent to bidding with no winner for the round.
- Timer position on screen is dependent on number of players to ensure best visuals no matter the player count.

### CODE REVIEW

File: Assets/Scripts/Gamestate/MatchController.cs
Comment (toberge) [inline Assets/Scripts/Gamestate/MatchController.cs:52]: GlobalHUDController and globalHUD?

**Review by toberge — COMMENTED** [review]
Just a naming suggestion, I haven't tested this (though it probably runs fine).

File: Assets/Scripts/Gamestate/MatchController.cs
Comment (Fueredoriku) [reply Assets/Scripts/Gamestate/MatchController.cs:52]: As we have have a HUDController, I wanted to build on that specific name for something functionally similar. But I don't really have a strong opinion regarding this, so I can change it before merge

Comment (toberge) [reply Assets/Scripts/Gamestate/MatchController.cs]: I think we should rename the HUDController to something more specific though, like PlayerHUDController or LocalHUDController

File: Assets/Scripts/Gamestate/MatchController.cs
Comment (TArchetype) [inline Assets/Scripts/Gamestate/MatchController.cs:52]: Since this component is on the same game object  consider adding a `[requireComponent()]` field and so referencing that through a `GetComponent()` call.

File: Assets/Scripts/Timers/Timer.cs
Comment (TArchetype) [inline Assets/Scripts/Timers/Timer.cs:38]: Small nitpick, but you can remove `System.Collections.Generic` and `System.Data` namespaces from this script.

File: Assets/Scripts/UI/GlobalHUDController.cs
Comment (TArchetype) [inline Assets/Scripts/UI/GlobalHUDController.cs:21]: The timer text is not positioned correctly for when we have one or two players:

image
image

Consider expanding the `Text  (TMP)` component to cover the entire screen, then instead of setting the anchored position, switch between top centered text and middle centered text options in the text itself.

File: Assets/Scripts/Timers/Timer.cs
Comment (TArchetype) [inline Assets/Scripts/Timers/Timer.cs:39]: Is there any need to Serialize the timer components if they are all controlled by a different script?

**Review by TArchetype — CHANGES_REQUESTED** [review]
This is a great feature to have, adds a certain level of pressure to the players and keeps things interesting. It would also be great in the future to expand on this feature by adding an end of round result page which visually shows how much resources players earned in this match and have.

However I do have a couple of comments before merging:

File: Assets/Scripts/Timers/Timer.cs
Comment (Fueredoriku) [reply Assets/Scripts/Timers/Timer.cs:38]: You should discuss that with the author @Stektpotet, as it's outside the scope of the changes I implemented in this PR

File: Assets/Scripts/Gamestate/MatchController.cs
Comment (Fueredoriku) [reply Assets/Scripts/Gamestate/MatchController.cs:52]: I disagree for multiple reasons:

- Since this is in the scope of MatchController, this would have to be a getComponentInchildren or have a refference to the HUD object child which defeat the purpose.
- The matchmanager might find itself in situations where it doesn't even need the global HUD (like being in the bidding-scene).
- GetComponent is nice to avoid when it makes sense. MatchController itself being a required component on the MatchManager itself already implies that this script is prefab specific, meaning people are unlikely to find the requiredComponent useful.

File: Assets/Scripts/UI/GlobalHUDController.cs
Comment (Fueredoriku) [reply Assets/Scripts/UI/GlobalHUDController.cs:21]: The text is displayed correctly with any amount of players as long as the aspect ratio is normal. I will make a commit to make it work with other aspect rations than 16:9 though.

**Review by toberge — CHANGES_REQUESTED** [review]
LGTM, as long as you fix the aspect ratio issue @TArchetype mentioned and also the brief moment where the timer is **-1:0-1** before the bidding scene loads, it's ready to merge.

**Review by toberge — APPROVED** [review]
Given that we will replace the timer with some other UI when the time is up, the -1:-01 issue is insignificant.

**Review by TArchetype — APPROVED** [review]
Timer position fixed, approved

**Pull request merged**
